### PR TITLE
fitter_3d: consume authored joint_limits as a rotation prior during mesh registration (#97)

### PR DESCRIPTION
**Pull request by khaoulaJel**
Addresses #97.

## What this does

`SMAL3DFitter` / `Stage` (`fitter_3d/`) was the one fitting path with no
joint-rotation prior at all; `joint_rot` was optimised completely
unconstrained. This adds the same joint_limits hinge prior already used by
the optimisation fitter (`smal_fitter/fitter.py`) and the single-view /
multi-view neural regressors, so all four fitting paths now respect authored
per-joint rotation limits consistently.

- Added pure function `_joint_limit_tensors_from_dd(dd, device)` in
  `fitter_3d/trainer.py`, building `(min_limits, max_limits, error)` from an
  already-loaded SMAL `.pkl` dict via `_ranges_from_joint_limits`. It takes
  the locally-loaded `dd` directly rather than going through `LimitPrior()` /
  `config.dd`, avoiding dependency on a global staying synchronized with
  `config.SMAL_FILE`.
- Falls back to wide-open `[-pi, pi]` limits when `joint_limits` is absent;
  existing models without authored limits behave exactly as before.
- Malformed `joint_limits` are deferred rather than raised at construction.
  `SMAL3DFitter` is built unconditionally before any stage's `loss_weights`
  are read, so failing during construction would break runs that do not even
  enable the loss.
- `Stage.forward()` gets a sixth loss term, gated by `w_limit`. The Stage
  default remains `0.0` for backward compatibility, while the shipped
  `fitter_3d/ants_cfg.yaml` enables the prior with `w_limit: 100.0` for
  models that provide authored joint limits.
- Raises clearly if `w_limit > 0` but limits are unavailable (malformed file
  or legacy hardcoded-body mode, which does not support `joint_limits`).
- Collapsed the two separate `SMAL_FILE` reads in `SMAL3DFitter.__init__`
  into one, as required by the issue.

## Effect on models with authored joint limits

A new OmniAnt model (`OmniAnt_25PCs_joint_limited.pkl`) with authored
joint limits was used to validate the prior on real registrations. The model
contains limits for 99 of 165 axes across 33 of 55 joints.

With `w_limit: 100.0`, the prior eliminates joint-limit violations while
preserving surface fitting quality. Violations were removed without a
measurable regression in the final registration metrics, showing that the
prior acts as an anatomical constraint rather than restricting valid fits.

Models without a `joint_limits` key continue to use the wide-open fallback
behaviour.

## Testing

New `tests/test_fitter3d_joint_limits.py`:

- Fast unit tests covering:
  - fallback behaviour,
  - authored limits landing in the correct `(N_POSE, 3)` slot,
  - malformed limits being deferred rather than raised,
  - hinge-loss behaviour (zero inside limits, positive correction outside),
  - `w_limit=0` being a true no-op (key absent from `loss_components`).

- One `@pytest.mark.slow` end-to-end test:
  - builds a synthetic tight limit,
  - runs the real `SMAL3DFitter -> Stage.forward() -> optimizer` path with
    `w_limit=0` and `w_limit=100`,
  - verifies that the constrained fit has fewer joint-limit violations than
    the unconstrained fit.

Full suite: 105 passed, 6 skipped, 0 failed.

## Additional fixes

- Updated `_safe_ranges` to catch both `ValueError` and `TypeError` from
  malformed `joint_limits`, preserving the deferred-validation behaviour.
- Removed a CUDA-nondeterministic chamfer assertion from the joint-limit
  tests. The remaining assertions directly test the intended behaviour:
  reduction of joint-limit violations.

- Fixed `tests/test_config_system.py`:
  `test_apply_smal_file_override_sets_smal_file_and_shape_family` only
  restored 2 of the approximately 9 globals modified by
  `apply_smal_file_override`, causing model state (`N_POSE`, `dd`,
  `joint_names`, etc.) to leak between tests in the same pytest process.

### CODE REVIEW

File: fitter_3d/ants_cfg.yaml [hunk 1]
```diff
@@ -18,6 +18,7 @@ stages:
           w_edge: 0.8
           w_normal: 0.02
           w_laplacian: 0.01
+          w_limit: 100.0   # joint rotation limit prior, off by default - requires 'joint_limits' in the model .pkl (see docs/joint_limits_user_guide.md)
```
Comment (FabianPlum) [inline fitter_3d/ants_cfg.yaml]: This line is active, not commented — it turns the prior on at weight 100 for the shipped ants registration config, while its own trailing comment says "off by default". The PR description says the same ("off by default (`0.0`) … existing registrations unchanged unless explicitly enabled").

It's a no-op only while the model has no `joint_limits`. Against `OmniAnt_25PCs_joint_limited.pkl` it's live.

I measured it and 100.0 is a good value (see the review summary), so I'd keep it — but let's make it deliberate rather than incidental: either comment the line out to match the docs, or leave it on and fix both the trailing comment and the PR description to say it ships enabled. My preference is the latter, since the measurement supports it.

File: docs/joint_limits_user_guide.md [hunk 2]
```diff
@@ -151,6 +151,7 @@ The `.pkl` stores **radians**: −90° appears as `-1.5708`, and unconstrained a
 ## 9. Use the limits
 
 - **Optimisation fitter:** point `config.SMAL_FILE` at your exported `.pkl` and run with the limit weight on (`w_limit > 0`). Nothing else to configure.
+- **3D mesh registration (`fitter_3d/`):** the same `w_limit` weight, set per-stage under that stage's `loss_weights:` block in your YAML config (see the commented example in `fitter_3d/ants_cfg.yaml`). Off by default (`0.0`); models without a `joint_limits` key behave exactly as before. This is the fourth consuming path, alongside the optimisation fitter and the single-/multi-view neural regressors below.
```
Comment (FabianPlum) [inline docs/joint_limits_user_guide.md]: "see the commented example in `fitter_3d/ants_cfg.yaml`" — it isn't commented, it's an active `w_limit: 100.0`. Pairs with the `ants_cfg.yaml` comment; whichever way that one goes, this sentence needs to match.

File: .gitignore [hunk 3]
```diff
@@ -14,7 +14,7 @@ __pycache__/
 
 # Temp results
 Fitter_RESULTS/
-fit3d_results
+fit3d_results_*/
```
Comment (FabianPlum) [inline .gitignore]: `fit3d_results` → `fit3d_results_*/` narrows the pattern rather than widening it: it no longer matches a directory named exactly `fit3d_results`, which is the `results_dir` `ants_cfg.yaml` itself sets and where the README's documented command writes. Verified on this branch — after a default run `git status` shows `?? fit3d_results/`, where master ignores it.

`fit3d_results*/` catches both. Also flagging scope: this isn't related to #97. Fine to keep if deliberate.

File: tests/test_fitter3d_joint_limits.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        assert not (lo <= baseline_rot <= hi), (
+            f"baseline (unconstrained) run landed inside the synthetic bound ({baseline_rot:.4f}); "
+            "test doesn't provoke a real conflict, pick a different target joint/axis"
+        )
+        assert lo <= constrained_rot <= hi, (
+            f"constrained run (w_limit=100) still violates the bound: {constrained_rot:.4f} not in {target_range}"
+        )
+        # The constraint should cost something -- if chamfer is identical, the
+        # prior isn't actually trading off against the surface-fit objective.
+        assert constrained_chamfer >= baseline_chamfer
```
Comment (FabianPlum) [inline tests/test_fitter3d_joint_limits.py]: This assertion is flaky, and more importantly it isn't measuring the property worth asserting. It fails 6 runs in 8 on the GPU here.

Same seed, same code, 5 repeats of each arm:

```
baseline    chamfer 2.7437e-04 ± 1.66e-06
constrained chamfer 2.7521e-04 ± 1.31e-06

run-to-run spread of the baseline alone : 5.0e-06
the difference this asserts on          : 8.4e-07   <- 6x smaller
```

`torch.manual_seed(0)` doesn't fix it — the non-determinism is in the CUDA reductions in the chamfer backward, not in the sampling. CI is safe (`-m "not slow"`), but a bare `pytest` — what `CLAUDE.md` documents as "run all tests" — fails about two times in three.

Beyond the noise: "the constraint must cost chamfer" isn't a quality signal. Chamfer can be excellent on a fully collapsed model, so it can't tell a good constrained fit from a bad one. And empirically the constraint doesn't cost chamfer on real data — it slightly helps.

The other two assertions are the real content and are rock solid, 5/5 and 5/5. I'd drop this one.

A stronger test, if you want one (happy for this to be a follow-up issue so the PR can merge):

- pose the model at a known θ* drawn inside a synthetic limit box, forward it, export as the target — `generate_random_parameters` / `export_mesh_to_obj` in `fitter_3d/pointcloud2smil/sample_smil_model.py` already do this
- fit back from rest, w_limit 0 vs 100
- assert **total hinge across all 162 axes is lower** with the prior on — a population statement, so it doesn't ride on the luck of one hand-picked axis — and **mean |θ_fit − θ\*| is no worse**, which can't be faked by collapse and is the actual claim the prior makes
- keep chamfer only as a loose blow-up guard (`chamfer_B < 1.5 × chamfer_A`)

One catch to design around: with a clean mesh target and exact correspondence, chamfer may recover θ* on its own and leave the prior nothing to contribute. Sampling the target as a decimated point cloud (or dropping the tarsi) makes the problem underdetermined the way real scans are, which is where the prior actually earns its place.

File: fitter_3d/trainer.py [hunk 5]
```diff
@@ -36,6 +37,39 @@ def get_meshes(verts, faces, device="cuda"):
...
+def _joint_limit_tensors_from_dd(dd, device):
+    """Build (min_limits, max_limits, error) tensors of shape (N_POSE, 3) from an
+    already-loaded SMAL .pkl dict. `error` is the deferred ValueError (or None) if
+    'joint_limits' was present but malformed - validation is never raised here,
+    only reported, so construction never crashes a consumer that doesn't use w_limit."""
+
+    def _safe_ranges(dd):
+        try:
+            return _ranges_from_joint_limits(dd), None
+        except ValueError as e:
```
Comment (FabianPlum) [inline fitter_3d/trainer.py]: Minor: `_ranges_from_joint_limits` can raise `TypeError` as well as `ValueError` — `np.asarray(x, dtype=np.float64)` on a dict raises `TypeError`, which escapes and defeats the deferral guarantee this helper is built around. `except (ValueError, TypeError)` closes it.

I checked the realistic malformations and the current code handles them all correctly — wrong shape, min > max, NaN, a string and a ragged list all defer properly — so this really is an edge case.

Separately, the PR description quotes "0.44 rad unconstrained → 0.038 rad constrained". The same test here gives +0.688 → −0.013. Same conclusion, but the specific numbers aren't reproducible across hardware; worth softening to the qualitative claim.

Anyway, overall, very good work, let's merge after this tiny cleanup!

**Review by FabianPlum — COMMENTED** [review]
Nice work — the mechanism is right and mirrors `smal_fitter/fitter.py` faithfully (same hinge, same `[3:]` root slice, same mean-over-all-entries normalisation). Deferring a malformed `joint_limits` to first use instead of raising at construction is the correct call, and the `test_config_system.py` global-leak fix is a real catch worth having found.

I've been able to validate this beyond what was possible when you wrote it. I briefly made a new OmniAnt model with a first set of plausible joint ranges for legs and mandibles (leaving head, thorax and abdomen alone for now) as an at least temporary new production model, `OmniAnt_25PCs_joint_limited.pkl`, which **does** carry authored `joint_limits`: 99 of 165 axes constrained across 33 of 55 joints, median half-width 20°. So the "no-op on the production model today" note in the description no longer holds — this PR goes live the moment that model is used.

**I've pushed that model to this branch**, so the limit loss can now be exercised against a real model rather than only the wide-open fallback — point `config.SMAL_FILE` at it.

I ran the A/B that enables, on 81 ALL_ANTS_CLEAN registrations (the corpus the OmniAnt shape space was built from), full `ants_cfg.yaml` schedule, identical seed, only `w_limit` differing:

| | w_limit=0 | w_limit=100 |
|---|---|---|
| total hinge violation | 0.078 rad | **0.000** |
| specimens violating | 8–10 of 20 | **0** |
| mandible residual (final) | 0.00366 | −1.2% … −1.9% |
| mandible p90 (final) | 0.00607 | −1.8% … −3.8% |
| global residual (final) | 0.00343 | −0.2% … −0.4% |

**The prior eliminates every violation at no surface cost — slightly better, if anything.** All violations sit on `ma_r`/`ma_l` axis 1, and the taxa involved are the trap-jaw and unusually long/short mandibled species, so it's tracking real anatomy rather than fitting noise. It's free because the other two mandible axes are wide open (±30°/±70°) and Stages 2–3 let `deform_verts` absorb the remainder.

So `w_limit: 100.0` is a well-chosen value. Four things before merge, only one behavioural.

File: fitter_3d/ants_cfg.yaml (hunk 1 above)
Comment (FabianPlum) [reply fitter_3d/ants_cfg.yaml]: Resolved in 60cc147 with the preferred option: the prior stays enabled at `w_limit: 100.0` (the measurement supports it), and the trailing comment now says so explicitly ("enabled - requires 'joint_limits' in the model .pkl"). The PR description was updated to match — it now states the shipped `ants_cfg.yaml` enables the prior rather than claiming off-by-default.

File: docs/joint_limits_user_guide.md (hunk 2 above)
Comment (FabianPlum) [reply docs/joint_limits_user_guide.md]: Resolved in 60cc147, matching the direction taken in the `ants_cfg.yaml` thread: the sentence now reads "see `fitter_3d/ants_cfg.yaml`, which ships with `w_limit: 100.0` enabled" — no more claim of a commented example.

File: .gitignore (hunk 3 above)
Comment (FabianPlum) [reply .gitignore]: Resolved in 60cc147 by reverting the change entirely — `.gitignore` on this branch is now byte-identical to master, so a directory named exactly `fit3d_results` (the default `results_dir`) is ignored again, and the out-of-scope edit is gone.

File: tests/test_fitter3d_joint_limits.py (hunk 4 above)
Comment (FabianPlum) [reply tests/test_fitter3d_joint_limits.py]: Resolved: the chamfer-cost assertion was dropped in 60cc147, keeping the two solid assertions (unconstrained run provokes a real violation; constrained run lands inside bounds). The docstrings still described the removed assertion, so 43d9144 cleans those up too. The stronger synthetic theta* round-trip test you sketched is filed as follow-up #99 so this PR can merge.

File: fitter_3d/trainer.py (hunk 5 above)
Comment (FabianPlum) [reply fitter_3d/trainer.py]: Resolved in 60cc147: `_safe_ranges` now catches `(ValueError, TypeError)`, so a dict-shaped `joint_limits` defers like every other malformation instead of escaping construction. The PR description was also softened to the qualitative claim (violations eliminated, no measurable regression) instead of quoting hardware-specific rad numbers.
